### ISSUE TITLE
Feature Request - More Code Formatting Authority / Options

### ISSUE THREAD

**Issue by berge459** [mathworks/MATLAB-extension-for-vscode#7]
It would be incredibly useful for teams to be able to enforce code style guidelines a la something like clang-format. Additionally, features like those provided in clang-tidy for variable/function/class naming conventions would help teams to produce more uniform MATLAB codebases with minimal manual effort. 

A simpler start to this could just include a "comfortable" vs. "compact" preference on auto formatting: 
Comfortable: `matlab_function(arg1, arg2, arg3);`
Compact: `matlab_function(arg1,arg2,arg3);`

 or 
Comfortable: `var = x * y;`
Compact: `var=x*y;`

**Comment by dklilley** [comment]
> Currently when I save a file it flattens everything within a function that isn't part of a code block to have no indentation.

@Whyhow314 - Document formatting currently listens to the preferences defined in MATLAB. If you open MATLAB, you should be able to modify the `Function indenting format` preference under "MATLAB > Editor/Debugger > Language". Perhaps the behavior with `Indent all functions` is closer to what you would prefer?

**Comment by Whyhow314** [comment]
@dklilley Looks like that fixed it, thanks! I didn't realize that it respected the preferences from my matlab installation.

**Comment by tongyaop** [comment]
Does this extension include auto indenting?

**Comment by tongyaop** [comment]
@dklilley Thank you for the reply. The code does indent when I type. However, I am looking for a command that is similar to `cmd + i` (or `control + i`) in MATLAB desktop that auto indents a piece of already-typed code. When pieces of code are copied from somewhere else, they are not always in the right indentation

**Comment by dklilley** [comment]
@tongyaop You can use Format Document for this. I believe the keyboard shortcut for this is `shift + option + F` (or `shift + alt + F`). There is not currently support in the extension for Format Selection, although this is something we hope to add in the future.

**Comment by dfkoza-vf** [comment]
We are also desperately waiting for auto-formatting functionality as mentioned by the original feature requestor (not just indentation, but code formatting - the same actually holds for the MATLAB IDE). Our dream would be a Matlab counterpart to Python's [black code formatter](https://github.com/psf/black).

It is often not software developers (but engineers or similar) that use Matlab, and hence auto-formatting of Matlab code can result in massive improvements of code quality, without the need to bother the contributing engineers with coding style rules.

@dklilley @nickchoi-mw  The request has been open since April, any estimate on when this will be released?

**Comment by dklilley** [comment]
This is an area we want to expand upon and are looking into as part of a multi-team effort. I am not able to provide any estimates at this time.

**Comment by yangtzech** [comment]
Is this feature still under consideration or dropped? A matlab version of Python's [black code formatter](https://github.com/psf/black) would be very much welcome. And there are some references in GitHub like [matlab-formatter-vscode](https://github.com/affenwiesel/matlab-formatter-vscode) or [matlab-formatter-plus](https://github.com/Redns/matlab-formatter-plus)

**Comment by dklilley** [comment]
Yes, this is something which we still want to bring to the extension. We are coordinating with other teams within MathWorks, but are unable to provide a timeline for when this will become available.
